### PR TITLE
fix: flag native keras config modules

### PR DESCRIPTION
**Pull request by mldangelo-oai**
## Summary

- classify exact native execution modules in Keras configs as critical CVE-2025-1550 findings
- inspect canonical string-config functions and nested serialized callable objects
- add `_xxsubinterpreters` while keeping unimportable dotted native-module children below critical severity
- preserve benign metadata dictionaries that merely contain a `module` key

## Critical review

The review reproduced two false negatives on the prior head: canonical `posix.system` function objects with string configs and nested serialized activation functions were not inspected. It also found that `_xxsubinterpreters.run_string` was omitted and that `posix.path` was incorrectly promoted even though native extension modules are not packages. The pushed head addresses all four cases with iterative traversal inside the existing bounded Keras config read.

## Validation

- CVE-2025-1550 focused tests: `46 passed`
- changed scanner test module: `250 passed, 38 skipped` (optional `h5py` unavailable)
- Ruff clean on the changed scanner and test files
- mypy clean on the changed scanner and test files
- `git diff --check` clean

The branch is synced with current `main`.

## Scan finding

- Fixes MA-DSS-C034 from the Codex Security scan.

### CODE REVIEW

**Comment by mldangelo-oai** [comment]
@codex review

**Review by ianw-oai — APPROVED** [review]

**Comment by mldangelo-oai** [comment]
Critical review found and fixed two exploitable coverage gaps: canonical Keras function objects with string `config` values were skipped, and nested serialized callables such as activations were never inspected. I also added the direct `_xxsubinterpreters.run_string` module path and stopped promoting impossible dotted children of native extension modules.

Validation on the pushed current-main head: focused CVE tests `46 passed`; full changed scanner module `250 passed, 38 skipped`; Ruff, mypy, and diff checks clean. Moving to the next PR while CI runs.

**Comment by mldangelo-oai** [comment]
QA complete on head `8176eeaf`.

Additional review fixes:
- classify Python 3.13 `_interpreters` as an exact dangerous native module, including malicious, dotted-child, and prefix-collision regressions
- require the canonical Keras serialized-object shape (`class_name` plus an explicit `config` key) before promoting layer or nested metadata module references
- preserve detection for string-config functions and nested serialized callables while removing malformed metadata false positives

Validation:
- `PROMPTFOO_DISABLE_TELEMETRY=1 .venv/bin/pytest tests/scanners/test_keras_zip_scanner.py::TestCVE20251550ModuleReferences -q -x` -> 52 passed
- `PROMPTFOO_DISABLE_TELEMETRY=1 .venv/bin/pytest tests/scanners/test_keras_zip_scanner.py -q -x` -> 294 passed
- scoped Ruff format/check -> clean
- scoped mypy -> clean
- `git diff --check` -> clean

The PR remains approved and has no review threads.

**Comment by mldangelo-oai** [comment]
Rebased the review result onto current `main` via merge and preserved both the newer Keras evidence/Lambda hardening and this PR’s native-module coverage. Additional QA found four exact CPython runtime-module gaps: `_imp` and `_frozen_importlib*` expose import/exec primitives, while `_io` exposes direct file access. Added them without broadening dotted descendants or prefix lookalikes.\n\nValidation on head `4e898da8`:\n- CVE-2025-1550 focused class: 64 passed\n- full changed Keras ZIP test module: 317 passed, 66 expected optional-dependency skips\n- scoped Ruff format/check: clean\n- scoped mypy: clean\n- `git diff --check`: clean\n\nNo unresolved review threads remain. Moving on while fresh CI runs.

File: modelaudit/scanners/keras_zip_scanner.py [hunk 1]
```diff
@@ -1588,93 +1612,153 @@ def _is_vulnerable_keras_3_11_x(version: str) -> bool | None:
...
+                self._check_config_module_reference(
+                    layer_config,
+                    key,
+                    config_value.strip(),
+                    layer_class,
+                    result,
+                    layer_name,
                 )
 
+        self._check_nested_serialized_module_references(layer_config, result, layer_name)
```
[the remark this answers is not in the document]
Comment (mldangelo-oai) [reply modelaudit/scanners/keras_zip_scanner.py]: Fixed in faa4ae94. The native-module traversal now covers trusted compile deserialization scopes for optimizer, loss, metrics, and weighted_metrics, including nested output mappings. Regressions verify malicious posix loss and nt metric functions are both critical.

**Comment by mldangelo-oai** [comment]
Fresh review on head faa4ae94 fixed both new threads. Native-module detection now covers compile-time optimizer/loss/metric deserialization, including nested output mappings, while nested layer traversal is limited to real Keras serialized-object surfaces so metadata dictionaries with class_name/module/config no longer become critical false positives. Validation: focused CVE class 65 passed; full Keras ZIP module 318 passed with 66 expected optional-dependency skips; scoped Ruff, format, mypy, and diff checks clean. All review threads are resolved. Moving on while CI runs.

File: modelaudit/scanners/keras_zip_scanner.py [hunk 2]
```diff
@@ -1588,93 +1666,172 @@ def _is_vulnerable_keras_3_11_x(version: str) -> bool | None:
...
```
[the remark this answers is not in the document]
Comment (mldangelo-oai) [reply modelaudit/scanners/keras_zip_scanner.py]: Fixed in ee44d538. The root serialized object is now checked before model/layer extraction, while arbitrary `model_config.config.module` metadata remains excluded to avoid a new false positive. Added positive root-callable and negative metadata regressions.

File: modelaudit/scanners/keras_zip_scanner.py [hunk 3]
```diff
@@ -1588,93 +1666,172 @@ def _is_vulnerable_keras_3_11_x(version: str) -> bool | None:
...
+                pending.extend((item, parent_key, container_is_trusted) for item in node)
+                continue
+            if not isinstance(node, dict):
+                continue
+
+            object_class = node.get("class_name")
+            serialized_shape = isinstance(object_class, str) and "config" in node
+            is_serialized_object = serialized_shape and (
+                container_is_trusted
+                or parent_key in _NESTED_SERIALIZED_OBJECT_KEYS
```
[the remark this answers is not in the document]
Comment (mldangelo-oai) [reply modelaudit/scanners/keras_zip_scanner.py:1829]: Fixed in ee44d538. `learning_rate` is now a trusted nested serialized-object key beneath optimizer configs, with a regression proving an `os.system` schedule is reported CRITICAL. The broader pass also added canonical inbound-node, FlaxLayer, and FeatureSpace traversal coverage.

**Comment by mldangelo-oai** [comment]
Review/QA update at `ee44d538`:

- fixed both open P1s: root serialized callables and optimizer `learning_rate` schedules
- added canonical traversal for Functional inbound-node args/kwargs, `FlaxLayer.config.module`, and FeatureSpace preprocessors hidden by unregistered wrappers
- replaced blanket native-module critical findings with module-plus-symbol rules, eliminating malformed `posix.Dense` false positives while covering aliases such as `io.open` / `_io.open` and `operator.attrgetter` / `_operator.attrgetter`
- preserved benign root metadata and inbound-node metadata near-matches
- replied to and resolved all review threads

Focused validation: CVE-2025-1550 class `89 passed`; full changed Keras ZIP test module `405 passed, 3 skipped` (optional ONNX unavailable); scoped Ruff, format, mypy, and `git diff --check` clean. The previous red CI signal was solely the 30-minute Quick Feedback timeout, not a test assertion failure.
